src/app/clusters/commodity-metering-server/commodity-metering-server.cpp

```diff
@@ -0,0 +1,360 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "commodity-metering-server.h"
+
+#include <protocols/interaction_model/StatusCode.h>
+
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/EventLogging.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Globals;
+using namespace chip::app::Clusters::Globals::Structs;
+using namespace chip::app::Clusters::CommodityMetering;
+using namespace chip::app::Clusters::CommodityMetering::Attributes;
+
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace CommodityMetering {
+
+// Some constraints for lists limitation
+constexpr uint8_t kMaxMeteredQuantityEntries                    = 128;
+constexpr uint8_t kMaxTariffComponentIDsPerMeteredQuantityEntry = 128;
+
+namespace {
+
+inline bool operator==(const Span<const uint32_t> & a, const Span<const uint32_t> & b)
+{
+    return std::equal(a.begin(), a.end(), b.begin());
+}
+
+inline bool operator==(const Structs::MeteredQuantityStruct::Type & lhs, const Structs::MeteredQuantityStruct::Type & rhs)
+{
+    return (lhs.tariffComponentIDs == rhs.tariffComponentIDs) && (lhs.quantity == rhs.quantity);
+}
+
+template <typename T>
+bool NullableListEqual(const DataModel::Nullable<DataModel::List<T>> & a, const DataModel::Nullable<DataModel::List<T>> & b)
+{
+    if (a.IsNull() || b.IsNull())
+    {
+        return a.IsNull() == b.IsNull();
+    }
+    else if (a.Value().size() == b.Value().size())
```

Comment:
Might be nicer to reverse the test and outdent, too:
```suggestion
    if (a.Value().size() != b.Value().size())
    {
        return false;
    }
    
    // Other logic.
```